src/main/scala/is/hail/expr/ir/Visit.scala

```diff
@@ -0,0 +1,64 @@
+package is.hail.expr.ir
+
+import is.hail.expr._
+import is.hail.expr.types._
+import is.hail.utils._
+import scala.collection.mutable._
+
+//
+// Traversing an IR tree/DAG - if a node has multiple
+// parents, it may be visited from each.
+//
+
+object Visit {
```

Comment:
I think of it as "visit nodes until escape", which can be used for something more general than Exists - e.g. it could do something isTotalCostGreaterThan(limit: Int): Boolean by escaping once it has seen enough nodes to hit the limit.  But I'll fit in with the scala terminology.

Reply:
Yes, that's exactly what Scala's `exists` function is supposed to be.